### PR TITLE
backport fix float16(::Union{Float64,BigFloat})

### PR DESCRIPTION
**Pull request by oscardssmith**

### CODE REVIEW

File: base/mpfr.jl [hunk 1]
```diff
@@ -337,9 +337,24 @@ Float64(x::BigFloat, r::RoundingMode) = Float64(x, convert(MPFRRoundingMode, r))
...
+    if (resi&0x7fffffff) < 0x38800000 # if Float16(res) is subnormal
+        #shift so that the mantissa lines up where it would for normal Float16
+        shift = 113-((resi & 0x7f800000)>>23)
+        if shift<23
+            resi |= 0x0080_0000 # set implicit bit
+            resi >>= shift
+        end
+    end
+    if (resi & 0x1fff == 0x1000) # if we are halfway between 2 Float16 values
+This conversation was marked as resolved by vchuravy
```
Comment (KristofferC) [inline base/mpfr.jl:352]: 👀

Comment (oscardssmith) [reply base/mpfr.jl:352]: oops

**Comment by KristofferC** [comment]
This seems to lack the tests from #42837, or?

**Comment by oscardssmith** [comment]
My thinking was that as a backport it is less important to have the test since people are very unlikely to change the implantation within 1.7, and this minimized size of the backport. Want me to add the tests?

**Comment by KristofferC** [comment]
To me it seems just as important, since we are introducing new code into an environment where the code hasn't been run before and want to make sure all the functionality is valid there. The tests are always backported together with bug fixes.

**Comment by oscardssmith** [comment]
tests added.

**Pull request merged**